freshmaker/logger.py

```diff
@@ -102,12 +101,12 @@ def setup_logger():
         backupCount=1,
         mode="a",
     )
-    file_handler.setLevel(LOG_LEVEL)
+    file_handler.setLevel(conf.log_level)
     file_handler.setFormatter(file_formatter)
     logger.addHandler(file_handler)
     if os.environ.get("CONSOLE_LOGS", "").lower() != "false":
```

Comment:
This adds another console stream handler:
```
logger.addHandler(console_handler)
```
that's why we see duplicate log lines (with different formats).

When the root logger is configured by:
```
logging.basicConfig(level=conf.log_level, format=log_format)
```
in `init_logging`, a console stream handler is enabled implicitly, so we have two console stream handlers enabled at the end.

Reply:
I see! Thanks for clarifying. I think I'm almost getting this solved, so I guess I'll try a bit more on this PR!

Indeed, the functions `init_logging` and `setup_logger` were instantiating two different loggers with their own handlers:
``` python
(init_logging)
<RootLogger root (INFO)>
[<StreamHandler <stderr> (NOTSET)>, <JournalHandler (NOTSET)>]

(setup_logger)
<Logger freshmaker (INFO)>
[<RotatingFileHandler /var/log/freshmaker/freshmaker.splunk.log (INFO)>, <StreamHandler <stdout> (INFO)>]
```
The `getLogger()` commands in `init_logging` were not referencing to the module/app name, and were therefore referencing to the root logger. I changed them to `logging.getLogger("freshmaker")`:
https://github.com/redhat-exd-rebuilds/freshmaker/blob/344423d7d76356c4e69ca16ab224f87188ba5266/freshmaker/logger.py#L146-L161
and now we only have one logger:
``` python
(init_logging)
<Logger freshmaker (INFO)>
[<JournalHandler (NOTSET)>]

(setup_logger)
<Logger freshmaker (INFO)>
[<JournalHandler (NOTSET)>, <RotatingFileHandler /var/log/freshmaker/freshmaker.splunk.log (INFO)>, <StreamHandler <stdout> (INFO)>]
```

The handler to stderr, which was previously attached to the root logger, is gone. But I think there wasn't a second console handler to stdout. (At least not for now, as the current deployments are using `log_backend=journal` if I'm not mistaken). Since the duplicated lines were not just for errors, I thought there must be something else to this issue.

I'm under the impression that as long as we have two handlers, one for the console and another for the log file, we will have those duplicated lines when we use `oc logs` (although they have different formatting). The solution that I could think for that was to set console logging conditionally to some configuration value, and in fact the logic for this was already in the code! 

I made an adjustment to the condition for setting console logging, and I introduced a new variable to the configuration files to control it. I set its default value to False, that way we can avoid logging to console:
https://github.com/redhat-exd-rebuilds/freshmaker/blob/344423d7d76356c4e69ca16ab224f87188ba5266/freshmaker/logger.py#L107-L108

(I believe the previous condition was not working as intended; if the "CONSOLE_LOGS" variable was not set, it would still make the console logging, since `("" != "false") = True` )

What do you think about these changes?

Thanks a lot again, @qixiang! You are always very helpful!

Reply:
The logging module is designed in some way against Python's zen, so it easily causes some confusions. A very important thing to understand this issue is these lines from https://docs.python.org/3/library/logging.html#logging.basicConfig
```
logging.basicConfig(**kwargs)
Does basic configuration for the logging system by creating a StreamHandler with
a default Formatter and adding it to the root logger. The functions debug(), info(),
warning(), error() and critical() will call basicConfig() automatically if no
handlers are defined for the root logger.

This function does nothing if the root logger already has handlers configured,
unless the keyword argument force is set to True.
```
So whenever you call `logging.basicConfig` (or call the `logging.debug` functions) it will **create a StreamHandler and add it to the root logger** unless the root handler is already has some handlers configured.

In our code:
```
logging.basicConfig(level=conf.log_level, format=log_format)
```
this configured the root logger with a stream handler attached, then if we have some code like:
```
import logging
logging.basicConfig(level=logging.ERROR, format="%(levelname)s: %(message)s")
freshmaker_log = logging.getLogger("freshmaker")
freshmaker_log.setLevel(logging.INFO)
freshmaker_log.info("test line")
```
the output is:
```
INFO: test line
```
The reason is `freshmaker_log` doesn't have any handlers, but the root logger has a stream handler (configured by `logging.basicConfig`, so the line is printed by the root logger.

In above example, another thing which can also be confusing is the root logger has log level set to `ERROR`, we called `freshmaker_log.info()`, but root logger can still print the line, this is because the log level is only checked by the `freshmaker_log` which the message was given to, and all other loggers doesn't check log level at all.

Reply:
Ah, I see! I was misunderstanding the issue again... but now I got it! Thanks for you patience, @qixiang!

I reverted this last attempt I made for not logging to the console, and instead implemented a function to remove the console handler of the root logger. I guess that would do the trick, right?